{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## This notebook handles the pre-processing needed to modify thereaction bounds in a cobra model to facilitate running Eflux2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append('../src')\n",
    "sys.path.append('/Users/mahs128/Repos/CONCERTO')\n",
    "from concerto.helpers.load_model_from_git import load_model_from_git\n",
    "from eflux2 import EFlux2\n",
    "import cobra\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import gurobipy\n",
    "from fba_utils import convert_transcriptomics_to_enzyme_activity"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. Create surrogate model for the reference strain by tightening bounds on fluxes\n",
    "\n",
    "#### 1a. Obtain optimized fluxes from the model constrained to the same criteria used to select the reference strain\n",
    "\n",
    "#### 1b. Set optimized fluxes of reactions of interest in the model upper bounds \n",
    "\n",
    "#### 1c. Run flux variability analysis to obtain (reasonably) tight bounds on the fluxes for all other reactions in the model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Set parameter TokenServer to value \"leghorn.emsl.pnl.gov\"\n"
     ]
    }
   ],
   "source": [
    "# Load SBML model\n",
    "# syn_model = cobra.io.read_sbml_model('../models/iJB785_w_sucrose_transport.xml')\n",
    "syn_model = cobra.io.read_sbml_model('../models/syn_elong.xml')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Maximize\n",
      "1.0*BIOMASS__1 - 1.0*BIOMASS__1_reverse_063c7\n"
     ]
    }
   ],
   "source": [
    "print(syn_model.objective)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fluxes</th>\n",
       "      <th>reduced_costs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>EX_gln__L_e</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_hco3_e</th>\n",
       "      <td>-1.990000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_mn2_e</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_arg__L_e</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADPT</th>\n",
       "      <td>0.000006</td>\n",
       "      <td>8.881784e-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZDS</th>\n",
       "      <td>0.000323</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RDXRr</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MPTSS</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NOR</th>\n",
       "      <td>0.000323</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CSND</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1090 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               fluxes  reduced_costs\n",
       "EX_gln__L_e  0.000000   0.000000e+00\n",
       "EX_hco3_e   -1.990000   0.000000e+00\n",
       "EX_mn2_e     0.000000   0.000000e+00\n",
       "EX_arg__L_e  0.000000   0.000000e+00\n",
       "ADPT         0.000006   8.881784e-16\n",
       "...               ...            ...\n",
       "ZDS          0.000323   0.000000e+00\n",
       "RDXRr        0.000000  -0.000000e+00\n",
       "MPTSS        0.000000   0.000000e+00\n",
       "NOR          0.000323   0.000000e+00\n",
       "CSND         0.000000   0.000000e+00\n",
       "\n",
       "[1090 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "fluxes           0.011\n",
       "reduced_costs    0.000\n",
       "Name: EX_sucr_e, dtype: float64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "fluxes          -0.263564\n",
       "reduced_costs    0.000000\n",
       "Name: EX_co2_e, dtype: float64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "fluxes           5.390524e-02\n",
       "reduced_costs    5.684342e-14\n",
       "Name: BIOMASS__1, dtype: float64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Run optimization on cobra model to get fluxes for reactions of interest\n",
    "# sucrose production ('EX_sucr_e') and CO2 uptake ('EX_co2_e') and biomass ('BIOMASS__1')\n",
    "opt_df = syn_model.optimize().to_frame()\n",
    "display(opt_df)\n",
    "display(opt_df.loc['EX_sucr_e'])\n",
    "display(opt_df.loc['EX_co2_e'])\n",
    "display(opt_df.loc['BIOMASS__1'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "EX_sucr_e :  0.011 ;   1000.0\n",
      "EX_co2_e :  -0.263564375437085 ;   1000.0\n",
      "BIOMASS__1 :  0.053905238809111765 ;   2.0\n"
     ]
    }
   ],
   "source": [
    "# Specify upper/lower bounds of model for excluded reactions\n",
    "for rxn in ['EX_sucr_e', 'EX_co2_e', 'BIOMASS__1']:\n",
    "   syn_model.reactions.get_by_id(rxn).lower_bound = opt_df.loc[rxn, 'fluxes']\n",
    "\n",
    "   print(rxn, \": \", syn_model.reactions.get_by_id(rxn).lower_bound, \";  \", syn_model.reactions.get_by_id(rxn).upper_bound)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create reactions list exclude reactions of interest: sucrose production ('EX_sucr_e') and CO2 uptake ('EX_co2_e') and biomass ('BIOMASS__1') \n",
    "rxn_list = [r.id for r in syn_model.reactions if (r.id not in ['EX_sucr_e', 'EX_co2_e', 'BIOMASS__1'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Set parameter TokenServer to value \"leghorn.emsl.pnl.gov\"\n",
      "Read LP format model from file /var/folders/k9/b8pxky2572sdtgy2vnxhcljw0000gn/T/tmpulf48yrc.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 928 rows, 2181 columns, 9211 nonzeros\n",
      "Set parameter TokenServer to value \"leghorn.emsl.pnl.gov\"\n",
      "Read LP format model from file /var/folders/k9/b8pxky2572sdtgy2vnxhcljw0000gn/T/tmpkpquzsrs.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 928 rows, 2181 columns, 9211 nonzeros\n",
      "Set parameter TokenServer to value \"leghorn.emsl.pnl.gov\"\n",
      "Read LP format model from file /var/folders/k9/b8pxky2572sdtgy2vnxhcljw0000gn/T/tmpk4g90c7e.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 928 rows, 2181 columns, 9211 nonzeros\n",
      "Set parameter TokenServer to value \"leghorn.emsl.pnl.gov\"\n",
      "Read LP format model from file /var/folders/k9/b8pxky2572sdtgy2vnxhcljw0000gn/T/tmpuabcugjy.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 928 rows, 2181 columns, 9211 nonzeros\n",
      "Set parameter TokenServer to value \"leghorn.emsl.pnl.gov\"\n",
      "Read LP format model from file /var/folders/k9/b8pxky2572sdtgy2vnxhcljw0000gn/T/tmpmvqcdgmn.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 928 rows, 2181 columns, 9211 nonzeros\n",
      "Set parameter TokenServer to value \"leghorn.emsl.pnl.gov\"\n",
      "Read LP format model from file /var/folders/k9/b8pxky2572sdtgy2vnxhcljw0000gn/T/tmpvkxfxgw1.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 928 rows, 2181 columns, 9211 nonzeros\n",
      "Set parameter TokenServer to value \"leghorn.emsl.pnl.gov\"\n",
      "Read LP format model from file /var/folders/k9/b8pxky2572sdtgy2vnxhcljw0000gn/T/tmpqul6fe0h.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 928 rows, 2181 columns, 9211 nonzeros\n",
      "Set parameter TokenServer to value \"leghorn.emsl.pnl.gov\"\n",
      "Read LP format model from file /var/folders/k9/b8pxky2572sdtgy2vnxhcljw0000gn/T/tmpjjmzcycv.lp\n",
      "Reading time = 0.01 seconds\n",
      ": 928 rows, 2181 columns, 9211 nonzeros\n",
      "Set parameter TokenServer to value \"leghorn.emsl.pnl.gov\"\n",
      "Read LP format model from file /var/folders/k9/b8pxky2572sdtgy2vnxhcljw0000gn/T/tmpqoju_wx1.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 928 rows, 2181 columns, 9211 nonzeros\n",
      "Set parameter TokenServer to value \"leghorn.emsl.pnl.gov\"\n",
      "Read LP format model from file /var/folders/k9/b8pxky2572sdtgy2vnxhcljw0000gn/T/tmpegv47wq6.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 928 rows, 2181 columns, 9211 nonzeros\n",
      "Set parameter TokenServer to value \"leghorn.emsl.pnl.gov\"\n",
      "Read LP format model from file /var/folders/k9/b8pxky2572sdtgy2vnxhcljw0000gn/T/tmpjznedx23.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 928 rows, 2181 columns, 9211 nonzeros\n",
      "Set parameter TokenServer to value \"leghorn.emsl.pnl.gov\"\n",
      "Read LP format model from file /var/folders/k9/b8pxky2572sdtgy2vnxhcljw0000gn/T/tmp_w96hfq1.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 928 rows, 2181 columns, 9211 nonzeros\n",
      "Set parameter TokenServer to value \"leghorn.emsl.pnl.gov\"\n",
      "Read LP format model from file /var/folders/k9/b8pxky2572sdtgy2vnxhcljw0000gn/T/tmpzenyyqb5.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 928 rows, 2181 columns, 9211 nonzeros\n",
      "Set parameter TokenServer to value \"leghorn.emsl.pnl.gov\"\n",
      "Read LP format model from file /var/folders/k9/b8pxky2572sdtgy2vnxhcljw0000gn/T/tmpagf91y8g.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 928 rows, 2181 columns, 9211 nonzeros\n",
      "Set parameter TokenServer to value \"leghorn.emsl.pnl.gov\"\n",
      "Read LP format model from file /var/folders/k9/b8pxky2572sdtgy2vnxhcljw0000gn/T/tmphfji3k5v.lp\n",
      "Reading time = 0.00 seconds\n",
      ": 928 rows, 2181 columns, 9211 nonzeros\n",
      "Set parameter TokenServer to value \"leghorn.emsl.pnl.gov\"\n",
      "Read LP format model from file /var/folders/k9/b8pxky2572sdtgy2vnxhcljw0000gn/T/tmpn6w6yzv0.lp\n",
      "Reading time = 0.01 seconds\n",
      ": 928 rows, 2181 columns, 9211 nonzeros\n"
     ]
    }
   ],
   "source": [
    "# Run FVA (~30-40 seconds)\n",
    "FVA_object = cobra.flux_analysis.flux_variability_analysis(model=syn_model, reaction_list=rxn_list, \n",
    "                                                           fraction_of_optimum=0.85, processes=8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>minimum</th>\n",
       "      <th>maximum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>EX_gln__L_e</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_hco3_e</th>\n",
       "      <td>-1.990000</td>\n",
       "      <td>-1.989450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_mn2_e</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_arg__L_e</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADPT</th>\n",
       "      <td>0.000006</td>\n",
       "      <td>0.000006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZDS</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RDXRr</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MPTSS</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NOR</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CSND</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1087 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              minimum   maximum\n",
       "EX_gln__L_e  0.000000  0.000000\n",
       "EX_hco3_e   -1.990000 -1.989450\n",
       "EX_mn2_e     0.000000  0.000000\n",
       "EX_arg__L_e  0.000000  0.000000\n",
       "ADPT         0.000006  0.000006\n",
       "...               ...       ...\n",
       "ZDS          0.000000  0.000339\n",
       "RDXRr        0.000000  0.000000\n",
       "MPTSS        0.000000  0.000000\n",
       "NOR          0.000000  0.000339\n",
       "CSND         0.000000  0.000000\n",
       "\n",
       "[1087 rows x 2 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "FVA_object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# of 0-flux reactions from FVA: 252\n",
      "Percent of fluxes in the model with of 0-values from FVA: 23.12 %\n"
     ]
    }
   ],
   "source": [
    "zero_flux_from_FVA = [i for i in FVA_object.index[(FVA_object.minimum==0) & (FVA_object.maximum==0)]]\n",
    "print(\"# of 0-flux reactions from FVA:\", len(zero_flux_from_FVA))\n",
    "print(\"Percent of fluxes in the model with of 0-values from FVA:\", round(100*len(zero_flux_from_FVA)/len(opt_df),2),\"%\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " ### 2. Next, convert the transcriptomics data to enzyme data (Take the min of subunits, take the sum of isozymes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Se_axen_d4_1</th>\n",
       "      <th>Se_axen_d4_2</th>\n",
       "      <th>Se_axen_d4_3</th>\n",
       "      <th>Se_axen_d6_1</th>\n",
       "      <th>Se_axen_d6_2</th>\n",
       "      <th>Se_axen_d6_3</th>\n",
       "      <th>Se_axen_d8_1</th>\n",
       "      <th>Se_axen_d8_2</th>\n",
       "      <th>Se_axen_d8_3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Label</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SYNPCC7942_RS00005</th>\n",
       "      <td>16290</td>\n",
       "      <td>17761</td>\n",
       "      <td>15101</td>\n",
       "      <td>14567</td>\n",
       "      <td>13967</td>\n",
       "      <td>12653</td>\n",
       "      <td>7016</td>\n",
       "      <td>8855</td>\n",
       "      <td>8689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SYNPCC7942_RS00010</th>\n",
       "      <td>7185</td>\n",
       "      <td>7502</td>\n",
       "      <td>6565</td>\n",
       "      <td>10086</td>\n",
       "      <td>7546</td>\n",
       "      <td>7705</td>\n",
       "      <td>3747</td>\n",
       "      <td>6670</td>\n",
       "      <td>6960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SYNPCC7942_RS00015</th>\n",
       "      <td>24176</td>\n",
       "      <td>26304</td>\n",
       "      <td>22781</td>\n",
       "      <td>23926</td>\n",
       "      <td>21306</td>\n",
       "      <td>20412</td>\n",
       "      <td>13440</td>\n",
       "      <td>17253</td>\n",
       "      <td>17053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SYNPCC7942_RS00020</th>\n",
       "      <td>35119</td>\n",
       "      <td>35145</td>\n",
       "      <td>25895</td>\n",
       "      <td>37701</td>\n",
       "      <td>34175</td>\n",
       "      <td>30569</td>\n",
       "      <td>24835</td>\n",
       "      <td>39769</td>\n",
       "      <td>30280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SYNPCC7942_RS00025</th>\n",
       "      <td>6891</td>\n",
       "      <td>7803</td>\n",
       "      <td>6607</td>\n",
       "      <td>6347</td>\n",
       "      <td>6284</td>\n",
       "      <td>6844</td>\n",
       "      <td>3326</td>\n",
       "      <td>4146</td>\n",
       "      <td>3180</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Se_axen_d4_1  Se_axen_d4_2  Se_axen_d4_3  Se_axen_d6_1  \\\n",
       "Label                                                                        \n",
       "SYNPCC7942_RS00005         16290         17761         15101         14567   \n",
       "SYNPCC7942_RS00010          7185          7502          6565         10086   \n",
       "SYNPCC7942_RS00015         24176         26304         22781         23926   \n",
       "SYNPCC7942_RS00020         35119         35145         25895         37701   \n",
       "SYNPCC7942_RS00025          6891          7803          6607          6347   \n",
       "\n",
       "                    Se_axen_d6_2  Se_axen_d6_3  Se_axen_d8_1  Se_axen_d8_2  \\\n",
       "Label                                                                        \n",
       "SYNPCC7942_RS00005         13967         12653          7016          8855   \n",
       "SYNPCC7942_RS00010          7546          7705          3747          6670   \n",
       "SYNPCC7942_RS00015         21306         20412         13440         17253   \n",
       "SYNPCC7942_RS00020         34175         30569         24835         39769   \n",
       "SYNPCC7942_RS00025          6284          6844          3326          4146   \n",
       "\n",
       "                    Se_axen_d8_3  \n",
       "Label                             \n",
       "SYNPCC7942_RS00005          8689  \n",
       "SYNPCC7942_RS00010          6960  \n",
       "SYNPCC7942_RS00015         17053  \n",
       "SYNPCC7942_RS00020         30280  \n",
       "SYNPCC7942_RS00025          3180  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load transciptomics data\n",
    "transcriptomics_fname = \"processed_data/cleaned_transcriptomics.csv\"\n",
    "transcriptomics_df = pd.read_csv(transcriptomics_fname, index_col=\"Label\")\n",
    "transcriptomics_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Se_axen_d8_3'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Set reference strain\n",
    "ref_rep = transcriptomics_df.columns[8] # 'Se_axen_d8_3'\n",
    "ref_rep"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Function to create dictionary of reactions to isozyme sets (corresponding genes from gene reaction rules)\n",
    "def get_gpr_dict(model):\n",
    "    \"\"\" Returns the gene reaction rule (GPR) for each reaction in the model.\"\"\"\n",
    "    # Parse GPR into a dict containing isozymes (separated by 'or')\n",
    "    # Each isozyme has a set of subunits (separated by 'and')\n",
    "    gpr_dict = dict()\n",
    "    for r in model.reactions:\n",
    "        if r.gene_reaction_rule:\n",
    "            isozymes = set()\n",
    "            for isozyme in [isozyme.strip('() ') for isozyme in r.gene_reaction_rule.split(' or ')]:\n",
    "                isozymes.add(frozenset(gene.strip('() ') for gene in isozyme.split(' and ')))\n",
    "            gpr_dict[r] = isozymes\n",
    "    \n",
    "    return gpr_dict\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "gpr_dict = get_gpr_dict(syn_model)\n",
    "display(gpr_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check which genes (listed in observed transcriptomics data) are missing from the cobra model\n",
    "# list_of_gprs_in_model = [str(syn_model.reactions.get_by_id(r.id).gpr).split(\" and \") for r in syn_model.reactions]\n",
    "list_of_gprs_in_model = [g.id for g in syn_model.genes]\n",
    "\n",
    "genes_not_in_model = [g for g in transcriptomics_df.index if g not in list_of_gprs_in_model]\n",
    "genes_not_in_data = [g for g in list_of_gprs_in_model if g not in transcriptomics_df.index]\n",
    "genes_in_model_and_data = [g for g in list_of_gprs_in_model if g in transcriptomics_df.index]\n",
    "\n",
    "\n",
    "print('# of genes in model:', len(list_of_gprs_in_model))\n",
    "print('# of genes in data:', len(transcriptomics_df.index))\n",
    "print('# of genes_not_in_model:', len(genes_not_in_model))\n",
    "print('# of genes_not_in_data:', len(genes_not_in_data))\n",
    "print('# of genes_in_model_and_data:', len(genes_in_model_and_data))\n",
    "\n",
    "print('# of reactions in model:', len([r.id for r in syn_model.reactions]))\n",
    "\n",
    "genes_not_in_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Code shared by Jeremy in Teams chat...\n",
    "from cobra import Reaction, Gene\n",
    "\n",
    "def gene_expression_to_enzyme_activity(gpr: dict[Reaction, list[list[Gene]]], expression: dict[Gene, float]):\n",
    "    \"\"\"Map gene expression to enzyme activity\n",
    "    inputs:\n",
    "        gpr: dictionary of reactions (keys) to list of list of genes (values) for the correpsonding gene reaction rule\n",
    "        expression: dictionary of gene names (keys) to values from [likely] observed transcriptomics data\n",
    "    outputs:\n",
    "        enzyme_activity: dictionary of reactions (keys) to corresponding isozyme activity from observed data (value)\n",
    "    \"\"\"\n",
    "    \n",
    "    enzyme_activity = {}\n",
    "    for rxn in syn_model.reactions:\n",
    "      # Initialize enzyme_activity for this reaction to 0-value\n",
    "      # Note: 0-value preserved IF this reaction doesn't have any genes in its gene reaction rule\n",
    "      # Obvious example: Exchange/transport reactions don't have corresponding genes in their reaction rule, so the 0-value is preserved\n",
    "      enzyme_activity[rxn] = 0.0\n",
    "      \n",
    "      if rxn in gpr: # ensure rxn has a gene_reaction_rule defined\n",
    "        for isozyme in gpr[rxn]:\n",
    "          # Initialize isozyme_activity for this isozyme to infinity\n",
    "          # Note: infinity-value is preserved IF this isozyme is not present in the observed transcriptomics data\n",
    "          isozyme_activity = np.inf\n",
    "          for gene in isozyme:\n",
    "\n",
    "            if gene in expression: # temporary fix: ensure gene is included in observed data\n",
    "              \n",
    "              isozyme_activity = np.min([isozyme_activity, expression[gene]])\n",
    "          enzyme_activity[rxn] += isozyme_activity\n",
    "    return enzyme_activity"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Function to convert transciptomics data to enzyme activity\n",
    "def convert_transcriptomics_to_enzyme_activity(transcriptomics_data: pd.DataFrame, gpr: dict[Reaction, list[list[Gene]]]):\n",
    "    \"\"\"Convert transcriptomics data to enzyme activity\n",
    "    inputs:\n",
    "        transcriptomics_data: dataframe of transcriptomics data\n",
    "        gpr: dictionary of reactions (keys) to list of list of genes (values) for the correpsonding gene reaction rule\n",
    "    outputs:\n",
    "        enzyme_activity_df: dataframe of enzyme activity converted from transcriptomics data\n",
    "    \"\"\"\n",
    "\n",
    "    # Initialize empty dataframe\n",
    "    enzyme_activity_df = pd.DataFrame()\n",
    "\n",
    "    # Loop through each strain to convert each column of transcriptomics data\n",
    "    for this_strain in transcriptomics_data.columns:\n",
    "        # Create dict of genes and corresponding float values using trancsciptomics data\n",
    "        expression_dict = {g: transcriptomics_data.loc[g][this_strain] for g in transcriptomics_data.index}\n",
    "        expr_dict_keys = [kz for kz in expression_dict.keys()]\n",
    "\n",
    "        # Run the gene expression to enzyme activity converter for this_strain\n",
    "        enzyme_activity_dict = gene_expression_to_enzyme_activity(gpr, expression_dict)\n",
    "\n",
    "        # Initialize empty dataframe \n",
    "        if this_strain == transcriptomics_data.columns[0]:\n",
    "            # Use enzyme_activity_dict keys as the index\n",
    "            enzyme_activity_df = enzyme_activity_df.reindex(enzyme_activity_dict.keys())\n",
    "            # Add reaction ID column\n",
    "            enzyme_activity_df['Reaction_ID'] = [k.id for k in enzyme_activity_dict.keys()]\n",
    "        \n",
    "        # Add enzymze_activity to dataframe\n",
    "        enzyme_activity_df[this_strain] = enzyme_activity_dict\n",
    "    \n",
    "    return enzyme_activity_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run enzyme activity converter for all strains in transcriptomics_df\n",
    "all_enzyme_activity_df = convert_transcriptomics_to_enzyme_activity(transcriptomics_df, syn_model)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Reaction_ID</th>\n",
       "      <th>Se_axen_d4_1</th>\n",
       "      <th>Se_axen_d4_2</th>\n",
       "      <th>Se_axen_d4_3</th>\n",
       "      <th>Se_axen_d6_1</th>\n",
       "      <th>Se_axen_d6_2</th>\n",
       "      <th>Se_axen_d6_3</th>\n",
       "      <th>Se_axen_d8_1</th>\n",
       "      <th>Se_axen_d8_2</th>\n",
       "      <th>Se_axen_d8_3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>EX_gln__L_e: gln__L_e --&gt;</th>\n",
       "      <td>EX_gln__L_e</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_hco3_e: hco3_e &lt;=&gt;</th>\n",
       "      <td>EX_hco3_e</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_mn2_e: mn2_e &lt;=&gt;</th>\n",
       "      <td>EX_mn2_e</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_arg__L_e: arg__L_e --&gt;</th>\n",
       "      <td>EX_arg__L_e</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADPT: ade_c + prpp_c &lt;=&gt; amp_c + ppi_c</th>\n",
       "      <td>ADPT</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O2tcx: o2_c --&gt; o2_cx</th>\n",
       "      <td>O2tcx</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AOXPBDC: 2a3pp_c + h_c --&gt; 3a2oxpp_c + co2_c</th>\n",
       "      <td>AOXPBDC</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DNTPPA: ahdt_c + h2o_c --&gt; dhpmp_c + h_c + ppi_c</th>\n",
       "      <td>DNTPPA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CBMD: cbm_c + 2.0 h_c --&gt; co2_c + nh4_c</th>\n",
       "      <td>CBMD</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BIOMASS_PIGMENTS: 0.1707 calxan_c + 0.1138 caro_c + 0.5691 cholphya_c + 0.5691 hpdcn_c + 0.0569 nstxan_c + 0.0524 phycy_c + 0.2276 zeax_c --&gt; bm_pigm_c</th>\n",
       "      <td>BIOMASS_PIGMENTS</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H2Otu_syn: h2o_u &lt;=&gt; h2o_c</th>\n",
       "      <td>H2Otu_syn</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DTMPK: atp_c + dtmp_c &lt;=&gt; adp_c + dtdp_c</th>\n",
       "      <td>DTMPK</td>\n",
       "      <td>1810.0</td>\n",
       "      <td>2030.0</td>\n",
       "      <td>1568.0</td>\n",
       "      <td>2020.0</td>\n",
       "      <td>1758.0</td>\n",
       "      <td>1736.0</td>\n",
       "      <td>946.0</td>\n",
       "      <td>1376.0</td>\n",
       "      <td>1598.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BIOMASS__1: 30.0 atp_c + 0.0977 bm_carbs_c + 0.00119 bm_cofactors_c + 0.0795 bm_cw_c + 0.0073 bm_dna_c + 0.159 bm_memlip_c + 0.0197 bm_pigm_c + 0.5112 bm_pro_c + 0.1136 bm_rna_c + 30.0 h2o_c --&gt; 30.0 adp_c + 30.0 h_c + 30.0 pi_c</th>\n",
       "      <td>BIOMASS__1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORNTA: akg_c + orn_c &lt;=&gt; glu5sa_c + glu__L_c</th>\n",
       "      <td>ORNTA</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATPM: atp_c + h2o_c --&gt; adp_c + h_c + pi_c</th>\n",
       "      <td>ATPM</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACOATA: ACP_c + accoa_c --&gt; acACP_c + coa_c</th>\n",
       "      <td>ACOATA</td>\n",
       "      <td>4691.0</td>\n",
       "      <td>5294.0</td>\n",
       "      <td>3826.0</td>\n",
       "      <td>4813.0</td>\n",
       "      <td>4716.0</td>\n",
       "      <td>4054.0</td>\n",
       "      <td>1790.0</td>\n",
       "      <td>3524.0</td>\n",
       "      <td>3105.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5DOAN: dad_5_c + h2o_c --&gt; 5drib_c + ade_c</th>\n",
       "      <td>5DOAN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3HAD160: 3hpalmACP_c &lt;=&gt; h2o_c + tpalm2eACP_c</th>\n",
       "      <td>3HAD160</td>\n",
       "      <td>1636.0</td>\n",
       "      <td>1826.0</td>\n",
       "      <td>1332.0</td>\n",
       "      <td>1591.0</td>\n",
       "      <td>1476.0</td>\n",
       "      <td>999.0</td>\n",
       "      <td>711.0</td>\n",
       "      <td>1132.0</td>\n",
       "      <td>1436.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNDR3: cdp_c + trdrd_c --&gt; dcdp_c + h2o_c + trdox_c</th>\n",
       "      <td>RNDR3</td>\n",
       "      <td>28195.0</td>\n",
       "      <td>26841.0</td>\n",
       "      <td>22287.0</td>\n",
       "      <td>19726.0</td>\n",
       "      <td>17817.0</td>\n",
       "      <td>15050.0</td>\n",
       "      <td>12523.0</td>\n",
       "      <td>10904.0</td>\n",
       "      <td>9458.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Htex: h_e &lt;=&gt; h_p</th>\n",
       "      <td>Htex</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G5SD: glu5p_c + h_c + nadph_c --&gt; glu5sa_c + nadp_c + pi_c</th>\n",
       "      <td>G5SD</td>\n",
       "      <td>20578.0</td>\n",
       "      <td>23123.0</td>\n",
       "      <td>19829.0</td>\n",
       "      <td>25519.0</td>\n",
       "      <td>20047.0</td>\n",
       "      <td>20535.0</td>\n",
       "      <td>12402.0</td>\n",
       "      <td>16955.0</td>\n",
       "      <td>19772.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HISTDa: histd_c + nad_c --&gt; h_c + histda_c + nadh_c</th>\n",
       "      <td>HISTDa</td>\n",
       "      <td>22946.0</td>\n",
       "      <td>25799.0</td>\n",
       "      <td>22763.0</td>\n",
       "      <td>29749.0</td>\n",
       "      <td>24686.0</td>\n",
       "      <td>26093.0</td>\n",
       "      <td>15009.0</td>\n",
       "      <td>23037.0</td>\n",
       "      <td>23684.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TRPS3: 3ig3p_c --&gt; g3p_c + indole_c</th>\n",
       "      <td>TRPS3</td>\n",
       "      <td>3922.0</td>\n",
       "      <td>4500.0</td>\n",
       "      <td>3382.0</td>\n",
       "      <td>4196.0</td>\n",
       "      <td>3584.0</td>\n",
       "      <td>3638.0</td>\n",
       "      <td>2308.0</td>\n",
       "      <td>3407.0</td>\n",
       "      <td>3780.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DMBZIDS2: fmnh2_c + h_c + nadh_c + o2_c --&gt; dialurate_c + dmbzid_c + e4p_c + h2o_c + nad_c</th>\n",
       "      <td>DMBZIDS2</td>\n",
       "      <td>11214.0</td>\n",
       "      <td>12125.0</td>\n",
       "      <td>10260.0</td>\n",
       "      <td>11519.0</td>\n",
       "      <td>10167.0</td>\n",
       "      <td>9387.0</td>\n",
       "      <td>5912.0</td>\n",
       "      <td>8199.0</td>\n",
       "      <td>7513.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSTPT: glu__L_c + imacp_c &lt;=&gt; akg_c + hisp_c</th>\n",
       "      <td>HSTPT</td>\n",
       "      <td>17080.0</td>\n",
       "      <td>17425.0</td>\n",
       "      <td>15619.0</td>\n",
       "      <td>17314.0</td>\n",
       "      <td>14256.0</td>\n",
       "      <td>13424.0</td>\n",
       "      <td>8566.0</td>\n",
       "      <td>11774.0</td>\n",
       "      <td>11645.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                         Reaction_ID  \\\n",
       "EX_gln__L_e: gln__L_e -->                                EX_gln__L_e   \n",
       "EX_hco3_e: hco3_e <=>                                      EX_hco3_e   \n",
       "EX_mn2_e: mn2_e <=>                                         EX_mn2_e   \n",
       "EX_arg__L_e: arg__L_e -->                                EX_arg__L_e   \n",
       "ADPT: ade_c + prpp_c <=> amp_c + ppi_c                          ADPT   \n",
       "O2tcx: o2_c --> o2_cx                                          O2tcx   \n",
       "AOXPBDC: 2a3pp_c + h_c --> 3a2oxpp_c + co2_c                 AOXPBDC   \n",
       "DNTPPA: ahdt_c + h2o_c --> dhpmp_c + h_c + ppi_c              DNTPPA   \n",
       "CBMD: cbm_c + 2.0 h_c --> co2_c + nh4_c                         CBMD   \n",
       "BIOMASS_PIGMENTS: 0.1707 calxan_c + 0.1138 caro...  BIOMASS_PIGMENTS   \n",
       "H2Otu_syn: h2o_u <=> h2o_c                                 H2Otu_syn   \n",
       "DTMPK: atp_c + dtmp_c <=> adp_c + dtdp_c                       DTMPK   \n",
       "BIOMASS__1: 30.0 atp_c + 0.0977 bm_carbs_c + 0....        BIOMASS__1   \n",
       "ORNTA: akg_c + orn_c <=> glu5sa_c + glu__L_c                   ORNTA   \n",
       "ATPM: atp_c + h2o_c --> adp_c + h_c + pi_c                      ATPM   \n",
       "ACOATA: ACP_c + accoa_c --> acACP_c + coa_c                   ACOATA   \n",
       "5DOAN: dad_5_c + h2o_c --> 5drib_c + ade_c                     5DOAN   \n",
       "3HAD160: 3hpalmACP_c <=> h2o_c + tpalm2eACP_c                3HAD160   \n",
       "RNDR3: cdp_c + trdrd_c --> dcdp_c + h2o_c + trd...             RNDR3   \n",
       "Htex: h_e <=> h_p                                               Htex   \n",
       "G5SD: glu5p_c + h_c + nadph_c --> glu5sa_c + na...              G5SD   \n",
       "HISTDa: histd_c + nad_c --> h_c + histda_c + na...            HISTDa   \n",
       "TRPS3: 3ig3p_c --> g3p_c + indole_c                            TRPS3   \n",
       "DMBZIDS2: fmnh2_c + h_c + nadh_c + o2_c --> dia...          DMBZIDS2   \n",
       "HSTPT: glu__L_c + imacp_c <=> akg_c + hisp_c                   HSTPT   \n",
       "\n",
       "                                                    Se_axen_d4_1  \\\n",
       "EX_gln__L_e: gln__L_e -->                                    0.0   \n",
       "EX_hco3_e: hco3_e <=>                                        0.0   \n",
       "EX_mn2_e: mn2_e <=>                                          0.0   \n",
       "EX_arg__L_e: arg__L_e -->                                    0.0   \n",
       "ADPT: ade_c + prpp_c <=> amp_c + ppi_c                       inf   \n",
       "O2tcx: o2_c --> o2_cx                                        0.0   \n",
       "AOXPBDC: 2a3pp_c + h_c --> 3a2oxpp_c + co2_c                 0.0   \n",
       "DNTPPA: ahdt_c + h2o_c --> dhpmp_c + h_c + ppi_c             0.0   \n",
       "CBMD: cbm_c + 2.0 h_c --> co2_c + nh4_c                      0.0   \n",
       "BIOMASS_PIGMENTS: 0.1707 calxan_c + 0.1138 caro...           0.0   \n",
       "H2Otu_syn: h2o_u <=> h2o_c                                   0.0   \n",
       "DTMPK: atp_c + dtmp_c <=> adp_c + dtdp_c                  1810.0   \n",
       "BIOMASS__1: 30.0 atp_c + 0.0977 bm_carbs_c + 0....           0.0   \n",
       "ORNTA: akg_c + orn_c <=> glu5sa_c + glu__L_c                 inf   \n",
       "ATPM: atp_c + h2o_c --> adp_c + h_c + pi_c                   0.0   \n",
       "ACOATA: ACP_c + accoa_c --> acACP_c + coa_c               4691.0   \n",
       "5DOAN: dad_5_c + h2o_c --> 5drib_c + ade_c                   0.0   \n",
       "3HAD160: 3hpalmACP_c <=> h2o_c + tpalm2eACP_c             1636.0   \n",
       "RNDR3: cdp_c + trdrd_c --> dcdp_c + h2o_c + trd...       28195.0   \n",
       "Htex: h_e <=> h_p                                            0.0   \n",
       "G5SD: glu5p_c + h_c + nadph_c --> glu5sa_c + na...       20578.0   \n",
       "HISTDa: histd_c + nad_c --> h_c + histda_c + na...       22946.0   \n",
       "TRPS3: 3ig3p_c --> g3p_c + indole_c                       3922.0   \n",
       "DMBZIDS2: fmnh2_c + h_c + nadh_c + o2_c --> dia...       11214.0   \n",
       "HSTPT: glu__L_c + imacp_c <=> akg_c + hisp_c             17080.0   \n",
       "\n",
       "                                                    Se_axen_d4_2  \\\n",
       "EX_gln__L_e: gln__L_e -->                                    0.0   \n",
       "EX_hco3_e: hco3_e <=>                                        0.0   \n",
       "EX_mn2_e: mn2_e <=>                                          0.0   \n",
       "EX_arg__L_e: arg__L_e -->                                    0.0   \n",
       "ADPT: ade_c + prpp_c <=> amp_c + ppi_c                       inf   \n",
       "O2tcx: o2_c --> o2_cx                                        0.0   \n",
       "AOXPBDC: 2a3pp_c + h_c --> 3a2oxpp_c + co2_c                 0.0   \n",
       "DNTPPA: ahdt_c + h2o_c --> dhpmp_c + h_c + ppi_c             0.0   \n",
       "CBMD: cbm_c + 2.0 h_c --> co2_c + nh4_c                      0.0   \n",
       "BIOMASS_PIGMENTS: 0.1707 calxan_c + 0.1138 caro...           0.0   \n",
       "H2Otu_syn: h2o_u <=> h2o_c                                   0.0   \n",
       "DTMPK: atp_c + dtmp_c <=> adp_c + dtdp_c                  2030.0   \n",
       "BIOMASS__1: 30.0 atp_c + 0.0977 bm_carbs_c + 0....           0.0   \n",
       "ORNTA: akg_c + orn_c <=> glu5sa_c + glu__L_c                 inf   \n",
       "ATPM: atp_c + h2o_c --> adp_c + h_c + pi_c                   0.0   \n",
       "ACOATA: ACP_c + accoa_c --> acACP_c + coa_c               5294.0   \n",
       "5DOAN: dad_5_c + h2o_c --> 5drib_c + ade_c                   0.0   \n",
       "3HAD160: 3hpalmACP_c <=> h2o_c + tpalm2eACP_c             1826.0   \n",
       "RNDR3: cdp_c + trdrd_c --> dcdp_c + h2o_c + trd...       26841.0   \n",
       "Htex: h_e <=> h_p                                            0.0   \n",
       "G5SD: glu5p_c + h_c + nadph_c --> glu5sa_c + na...       23123.0   \n",
       "HISTDa: histd_c + nad_c --> h_c + histda_c + na...       25799.0   \n",
       "TRPS3: 3ig3p_c --> g3p_c + indole_c                       4500.0   \n",
       "DMBZIDS2: fmnh2_c + h_c + nadh_c + o2_c --> dia...       12125.0   \n",
       "HSTPT: glu__L_c + imacp_c <=> akg_c + hisp_c             17425.0   \n",
       "\n",
       "                                                    Se_axen_d4_3  \\\n",
       "EX_gln__L_e: gln__L_e -->                                    0.0   \n",
       "EX_hco3_e: hco3_e <=>                                        0.0   \n",
       "EX_mn2_e: mn2_e <=>                                          0.0   \n",
       "EX_arg__L_e: arg__L_e -->                                    0.0   \n",
       "ADPT: ade_c + prpp_c <=> amp_c + ppi_c                       inf   \n",
       "O2tcx: o2_c --> o2_cx                                        0.0   \n",
       "AOXPBDC: 2a3pp_c + h_c --> 3a2oxpp_c + co2_c                 0.0   \n",
       "DNTPPA: ahdt_c + h2o_c --> dhpmp_c + h_c + ppi_c             0.0   \n",
       "CBMD: cbm_c + 2.0 h_c --> co2_c + nh4_c                      0.0   \n",
       "BIOMASS_PIGMENTS: 0.1707 calxan_c + 0.1138 caro...           0.0   \n",
       "H2Otu_syn: h2o_u <=> h2o_c                                   0.0   \n",
       "DTMPK: atp_c + dtmp_c <=> adp_c + dtdp_c                  1568.0   \n",
       "BIOMASS__1: 30.0 atp_c + 0.0977 bm_carbs_c + 0....           0.0   \n",
       "ORNTA: akg_c + orn_c <=> glu5sa_c + glu__L_c                 inf   \n",
       "ATPM: atp_c + h2o_c --> adp_c + h_c + pi_c                   0.0   \n",
       "ACOATA: ACP_c + accoa_c --> acACP_c + coa_c               3826.0   \n",
       "5DOAN: dad_5_c + h2o_c --> 5drib_c + ade_c                   0.0   \n",
       "3HAD160: 3hpalmACP_c <=> h2o_c + tpalm2eACP_c             1332.0   \n",
       "RNDR3: cdp_c + trdrd_c --> dcdp_c + h2o_c + trd...       22287.0   \n",
       "Htex: h_e <=> h_p                                            0.0   \n",
       "G5SD: glu5p_c + h_c + nadph_c --> glu5sa_c + na...       19829.0   \n",
       "HISTDa: histd_c + nad_c --> h_c + histda_c + na...       22763.0   \n",
       "TRPS3: 3ig3p_c --> g3p_c + indole_c                       3382.0   \n",
       "DMBZIDS2: fmnh2_c + h_c + nadh_c + o2_c --> dia...       10260.0   \n",
       "HSTPT: glu__L_c + imacp_c <=> akg_c + hisp_c             15619.0   \n",
       "\n",
       "                                                    Se_axen_d6_1  \\\n",
       "EX_gln__L_e: gln__L_e -->                                    0.0   \n",
       "EX_hco3_e: hco3_e <=>                                        0.0   \n",
       "EX_mn2_e: mn2_e <=>                                          0.0   \n",
       "EX_arg__L_e: arg__L_e -->                                    0.0   \n",
       "ADPT: ade_c + prpp_c <=> amp_c + ppi_c                       inf   \n",
       "O2tcx: o2_c --> o2_cx                                        0.0   \n",
       "AOXPBDC: 2a3pp_c + h_c --> 3a2oxpp_c + co2_c                 0.0   \n",
       "DNTPPA: ahdt_c + h2o_c --> dhpmp_c + h_c + ppi_c             0.0   \n",
       "CBMD: cbm_c + 2.0 h_c --> co2_c + nh4_c                      0.0   \n",
       "BIOMASS_PIGMENTS: 0.1707 calxan_c + 0.1138 caro...           0.0   \n",
       "H2Otu_syn: h2o_u <=> h2o_c                                   0.0   \n",
       "DTMPK: atp_c + dtmp_c <=> adp_c + dtdp_c                  2020.0   \n",
       "BIOMASS__1: 30.0 atp_c + 0.0977 bm_carbs_c + 0....           0.0   \n",
       "ORNTA: akg_c + orn_c <=> glu5sa_c + glu__L_c                 inf   \n",
       "ATPM: atp_c + h2o_c --> adp_c + h_c + pi_c                   0.0   \n",
       "ACOATA: ACP_c + accoa_c --> acACP_c + coa_c               4813.0   \n",
       "5DOAN: dad_5_c + h2o_c --> 5drib_c + ade_c                   0.0   \n",
       "3HAD160: 3hpalmACP_c <=> h2o_c + tpalm2eACP_c             1591.0   \n",
       "RNDR3: cdp_c + trdrd_c --> dcdp_c + h2o_c + trd...       19726.0   \n",
       "Htex: h_e <=> h_p                                            0.0   \n",
       "G5SD: glu5p_c + h_c + nadph_c --> glu5sa_c + na...       25519.0   \n",
       "HISTDa: histd_c + nad_c --> h_c + histda_c + na...       29749.0   \n",
       "TRPS3: 3ig3p_c --> g3p_c + indole_c                       4196.0   \n",
       "DMBZIDS2: fmnh2_c + h_c + nadh_c + o2_c --> dia...       11519.0   \n",
       "HSTPT: glu__L_c + imacp_c <=> akg_c + hisp_c             17314.0   \n",
       "\n",
       "                                                    Se_axen_d6_2  \\\n",
       "EX_gln__L_e: gln__L_e -->                                    0.0   \n",
       "EX_hco3_e: hco3_e <=>                                        0.0   \n",
       "EX_mn2_e: mn2_e <=>                                          0.0   \n",
       "EX_arg__L_e: arg__L_e -->                                    0.0   \n",
       "ADPT: ade_c + prpp_c <=> amp_c + ppi_c                       inf   \n",
       "O2tcx: o2_c --> o2_cx                                        0.0   \n",
       "AOXPBDC: 2a3pp_c + h_c --> 3a2oxpp_c + co2_c                 0.0   \n",
       "DNTPPA: ahdt_c + h2o_c --> dhpmp_c + h_c + ppi_c             0.0   \n",
       "CBMD: cbm_c + 2.0 h_c --> co2_c + nh4_c                      0.0   \n",
       "BIOMASS_PIGMENTS: 0.1707 calxan_c + 0.1138 caro...           0.0   \n",
       "H2Otu_syn: h2o_u <=> h2o_c                                   0.0   \n",
       "DTMPK: atp_c + dtmp_c <=> adp_c + dtdp_c                  1758.0   \n",
       "BIOMASS__1: 30.0 atp_c + 0.0977 bm_carbs_c + 0....           0.0   \n",
       "ORNTA: akg_c + orn_c <=> glu5sa_c + glu__L_c                 inf   \n",
       "ATPM: atp_c + h2o_c --> adp_c + h_c + pi_c                   0.0   \n",
       "ACOATA: ACP_c + accoa_c --> acACP_c + coa_c               4716.0   \n",
       "5DOAN: dad_5_c + h2o_c --> 5drib_c + ade_c                   0.0   \n",
       "3HAD160: 3hpalmACP_c <=> h2o_c + tpalm2eACP_c             1476.0   \n",
       "RNDR3: cdp_c + trdrd_c --> dcdp_c + h2o_c + trd...       17817.0   \n",
       "Htex: h_e <=> h_p                                            0.0   \n",
       "G5SD: glu5p_c + h_c + nadph_c --> glu5sa_c + na...       20047.0   \n",
       "HISTDa: histd_c + nad_c --> h_c + histda_c + na...       24686.0   \n",
       "TRPS3: 3ig3p_c --> g3p_c + indole_c                       3584.0   \n",
       "DMBZIDS2: fmnh2_c + h_c + nadh_c + o2_c --> dia...       10167.0   \n",
       "HSTPT: glu__L_c + imacp_c <=> akg_c + hisp_c             14256.0   \n",
       "\n",
       "                                                    Se_axen_d6_3  \\\n",
       "EX_gln__L_e: gln__L_e -->                                    0.0   \n",
       "EX_hco3_e: hco3_e <=>                                        0.0   \n",
       "EX_mn2_e: mn2_e <=>                                          0.0   \n",
       "EX_arg__L_e: arg__L_e -->                                    0.0   \n",
       "ADPT: ade_c + prpp_c <=> amp_c + ppi_c                       inf   \n",
       "O2tcx: o2_c --> o2_cx                                        0.0   \n",
       "AOXPBDC: 2a3pp_c + h_c --> 3a2oxpp_c + co2_c                 0.0   \n",
       "DNTPPA: ahdt_c + h2o_c --> dhpmp_c + h_c + ppi_c             0.0   \n",
       "CBMD: cbm_c + 2.0 h_c --> co2_c + nh4_c                      0.0   \n",
       "BIOMASS_PIGMENTS: 0.1707 calxan_c + 0.1138 caro...           0.0   \n",
       "H2Otu_syn: h2o_u <=> h2o_c                                   0.0   \n",
       "DTMPK: atp_c + dtmp_c <=> adp_c + dtdp_c                  1736.0   \n",
       "BIOMASS__1: 30.0 atp_c + 0.0977 bm_carbs_c + 0....           0.0   \n",
       "ORNTA: akg_c + orn_c <=> glu5sa_c + glu__L_c                 inf   \n",
       "ATPM: atp_c + h2o_c --> adp_c + h_c + pi_c                   0.0   \n",
       "ACOATA: ACP_c + accoa_c --> acACP_c + coa_c               4054.0   \n",
       "5DOAN: dad_5_c + h2o_c --> 5drib_c + ade_c                   0.0   \n",
       "3HAD160: 3hpalmACP_c <=> h2o_c + tpalm2eACP_c              999.0   \n",
       "RNDR3: cdp_c + trdrd_c --> dcdp_c + h2o_c + trd...       15050.0   \n",
       "Htex: h_e <=> h_p                                            0.0   \n",
       "G5SD: glu5p_c + h_c + nadph_c --> glu5sa_c + na...       20535.0   \n",
       "HISTDa: histd_c + nad_c --> h_c + histda_c + na...       26093.0   \n",
       "TRPS3: 3ig3p_c --> g3p_c + indole_c                       3638.0   \n",
       "DMBZIDS2: fmnh2_c + h_c + nadh_c + o2_c --> dia...        9387.0   \n",
       "HSTPT: glu__L_c + imacp_c <=> akg_c + hisp_c             13424.0   \n",
       "\n",
       "                                                    Se_axen_d8_1  \\\n",
       "EX_gln__L_e: gln__L_e -->                                    0.0   \n",
       "EX_hco3_e: hco3_e <=>                                        0.0   \n",
       "EX_mn2_e: mn2_e <=>                                          0.0   \n",
       "EX_arg__L_e: arg__L_e -->                                    0.0   \n",
       "ADPT: ade_c + prpp_c <=> amp_c + ppi_c                       inf   \n",
       "O2tcx: o2_c --> o2_cx                                        0.0   \n",
       "AOXPBDC: 2a3pp_c + h_c --> 3a2oxpp_c + co2_c                 0.0   \n",
       "DNTPPA: ahdt_c + h2o_c --> dhpmp_c + h_c + ppi_c             0.0   \n",
       "CBMD: cbm_c + 2.0 h_c --> co2_c + nh4_c                      0.0   \n",
       "BIOMASS_PIGMENTS: 0.1707 calxan_c + 0.1138 caro...           0.0   \n",
       "H2Otu_syn: h2o_u <=> h2o_c                                   0.0   \n",
       "DTMPK: atp_c + dtmp_c <=> adp_c + dtdp_c                   946.0   \n",
       "BIOMASS__1: 30.0 atp_c + 0.0977 bm_carbs_c + 0....           0.0   \n",
       "ORNTA: akg_c + orn_c <=> glu5sa_c + glu__L_c                 inf   \n",
       "ATPM: atp_c + h2o_c --> adp_c + h_c + pi_c                   0.0   \n",
       "ACOATA: ACP_c + accoa_c --> acACP_c + coa_c               1790.0   \n",
       "5DOAN: dad_5_c + h2o_c --> 5drib_c + ade_c                   0.0   \n",
       "3HAD160: 3hpalmACP_c <=> h2o_c + tpalm2eACP_c              711.0   \n",
       "RNDR3: cdp_c + trdrd_c --> dcdp_c + h2o_c + trd...       12523.0   \n",
       "Htex: h_e <=> h_p                                            0.0   \n",
       "G5SD: glu5p_c + h_c + nadph_c --> glu5sa_c + na...       12402.0   \n",
       "HISTDa: histd_c + nad_c --> h_c + histda_c + na...       15009.0   \n",
       "TRPS3: 3ig3p_c --> g3p_c + indole_c                       2308.0   \n",
       "DMBZIDS2: fmnh2_c + h_c + nadh_c + o2_c --> dia...        5912.0   \n",
       "HSTPT: glu__L_c + imacp_c <=> akg_c + hisp_c              8566.0   \n",
       "\n",
       "                                                    Se_axen_d8_2  Se_axen_d8_3  \n",
       "EX_gln__L_e: gln__L_e -->                                    0.0           0.0  \n",
       "EX_hco3_e: hco3_e <=>                                        0.0           0.0  \n",
       "EX_mn2_e: mn2_e <=>                                          0.0           0.0  \n",
       "EX_arg__L_e: arg__L_e -->                                    0.0           0.0  \n",
       "ADPT: ade_c + prpp_c <=> amp_c + ppi_c                       inf           inf  \n",
       "O2tcx: o2_c --> o2_cx                                        0.0           0.0  \n",
       "AOXPBDC: 2a3pp_c + h_c --> 3a2oxpp_c + co2_c                 0.0           0.0  \n",
       "DNTPPA: ahdt_c + h2o_c --> dhpmp_c + h_c + ppi_c             0.0           0.0  \n",
       "CBMD: cbm_c + 2.0 h_c --> co2_c + nh4_c                      0.0           0.0  \n",
       "BIOMASS_PIGMENTS: 0.1707 calxan_c + 0.1138 caro...           0.0           0.0  \n",
       "H2Otu_syn: h2o_u <=> h2o_c                                   0.0           0.0  \n",
       "DTMPK: atp_c + dtmp_c <=> adp_c + dtdp_c                  1376.0        1598.0  \n",
       "BIOMASS__1: 30.0 atp_c + 0.0977 bm_carbs_c + 0....           0.0           0.0  \n",
       "ORNTA: akg_c + orn_c <=> glu5sa_c + glu__L_c                 inf           inf  \n",
       "ATPM: atp_c + h2o_c --> adp_c + h_c + pi_c                   0.0           0.0  \n",
       "ACOATA: ACP_c + accoa_c --> acACP_c + coa_c               3524.0        3105.0  \n",
       "5DOAN: dad_5_c + h2o_c --> 5drib_c + ade_c                   0.0           0.0  \n",
       "3HAD160: 3hpalmACP_c <=> h2o_c + tpalm2eACP_c             1132.0        1436.0  \n",
       "RNDR3: cdp_c + trdrd_c --> dcdp_c + h2o_c + trd...       10904.0        9458.0  \n",
       "Htex: h_e <=> h_p                                            0.0           0.0  \n",
       "G5SD: glu5p_c + h_c + nadph_c --> glu5sa_c + na...       16955.0       19772.0  \n",
       "HISTDa: histd_c + nad_c --> h_c + histda_c + na...       23037.0       23684.0  \n",
       "TRPS3: 3ig3p_c --> g3p_c + indole_c                       3407.0        3780.0  \n",
       "DMBZIDS2: fmnh2_c + h_c + nadh_c + o2_c --> dia...        8199.0        7513.0  \n",
       "HSTPT: glu__L_c + imacp_c <=> akg_c + hisp_c             11774.0       11645.0  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Write dataframe to csv\n",
    "all_enzyme_activity_df.to_csv('processed_data/enzymze_activity.csv')\n",
    "\n",
    "all_enzyme_activity_df.iloc[0:25]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " ### 3. For each condition, normalize enzyme data with respect to the reference strain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Se_axen_d4_1</th>\n",
       "      <th>Se_axen_d4_2</th>\n",
       "      <th>Se_axen_d4_3</th>\n",
       "      <th>Se_axen_d6_1</th>\n",
       "      <th>Se_axen_d6_2</th>\n",
       "      <th>Se_axen_d6_3</th>\n",
       "      <th>Se_axen_d8_1</th>\n",
       "      <th>Se_axen_d8_2</th>\n",
       "      <th>Se_axen_d8_3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Reaction_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>EX_gln__L_e</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_hco3_e</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_mn2_e</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_arg__L_e</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADPT</th>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O2tcx</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AOXPBDC</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DNTPPA</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CBMD</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BIOMASS_PIGMENTS</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H2Otu_syn</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DTMPK</th>\n",
       "      <td>1.132666</td>\n",
       "      <td>1.270338</td>\n",
       "      <td>0.981227</td>\n",
       "      <td>1.264080</td>\n",
       "      <td>1.100125</td>\n",
       "      <td>1.086358</td>\n",
       "      <td>0.591990</td>\n",
       "      <td>0.861076</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BIOMASS__1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORNTA</th>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATPM</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACOATA</th>\n",
       "      <td>1.510789</td>\n",
       "      <td>1.704992</td>\n",
       "      <td>1.232206</td>\n",
       "      <td>1.550081</td>\n",
       "      <td>1.518841</td>\n",
       "      <td>1.305636</td>\n",
       "      <td>0.576490</td>\n",
       "      <td>1.134944</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5DOAN</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3HAD160</th>\n",
       "      <td>1.139276</td>\n",
       "      <td>1.271588</td>\n",
       "      <td>0.927577</td>\n",
       "      <td>1.107939</td>\n",
       "      <td>1.027855</td>\n",
       "      <td>0.695682</td>\n",
       "      <td>0.495125</td>\n",
       "      <td>0.788301</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNDR3</th>\n",
       "      <td>2.981074</td>\n",
       "      <td>2.837915</td>\n",
       "      <td>2.356418</td>\n",
       "      <td>2.085642</td>\n",
       "      <td>1.883802</td>\n",
       "      <td>1.591246</td>\n",
       "      <td>1.324064</td>\n",
       "      <td>1.152886</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Htex</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G5SD</th>\n",
       "      <td>1.040765</td>\n",
       "      <td>1.169482</td>\n",
       "      <td>1.002883</td>\n",
       "      <td>1.290664</td>\n",
       "      <td>1.013909</td>\n",
       "      <td>1.038590</td>\n",
       "      <td>0.627251</td>\n",
       "      <td>0.857526</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HISTDa</th>\n",
       "      <td>0.968840</td>\n",
       "      <td>1.089301</td>\n",
       "      <td>0.961113</td>\n",
       "      <td>1.256080</td>\n",
       "      <td>1.042307</td>\n",
       "      <td>1.101714</td>\n",
       "      <td>0.633719</td>\n",
       "      <td>0.972682</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TRPS3</th>\n",
       "      <td>1.037566</td>\n",
       "      <td>1.190476</td>\n",
       "      <td>0.894709</td>\n",
       "      <td>1.110053</td>\n",
       "      <td>0.948148</td>\n",
       "      <td>0.962434</td>\n",
       "      <td>0.610582</td>\n",
       "      <td>0.901323</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DMBZIDS2</th>\n",
       "      <td>1.492613</td>\n",
       "      <td>1.613869</td>\n",
       "      <td>1.365633</td>\n",
       "      <td>1.533209</td>\n",
       "      <td>1.353254</td>\n",
       "      <td>1.249434</td>\n",
       "      <td>0.786903</td>\n",
       "      <td>1.091308</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSTPT</th>\n",
       "      <td>1.466724</td>\n",
       "      <td>1.496350</td>\n",
       "      <td>1.341262</td>\n",
       "      <td>1.486818</td>\n",
       "      <td>1.224216</td>\n",
       "      <td>1.152769</td>\n",
       "      <td>0.735595</td>\n",
       "      <td>1.011078</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Se_axen_d4_1  Se_axen_d4_2  Se_axen_d4_3  Se_axen_d6_1  \\\n",
       "Reaction_ID                                                                \n",
       "EX_gln__L_e           0.000000      0.000000      0.000000      0.000000   \n",
       "EX_hco3_e             0.000000      0.000000      0.000000      0.000000   \n",
       "EX_mn2_e              0.000000      0.000000      0.000000      0.000000   \n",
       "EX_arg__L_e           0.000000      0.000000      0.000000      0.000000   \n",
       "ADPT                       inf           inf           inf           inf   \n",
       "O2tcx                 0.000000      0.000000      0.000000      0.000000   \n",
       "AOXPBDC               0.000000      0.000000      0.000000      0.000000   \n",
       "DNTPPA                0.000000      0.000000      0.000000      0.000000   \n",
       "CBMD                  0.000000      0.000000      0.000000      0.000000   \n",
       "BIOMASS_PIGMENTS      0.000000      0.000000      0.000000      0.000000   \n",
       "H2Otu_syn             0.000000      0.000000      0.000000      0.000000   \n",
       "DTMPK                 1.132666      1.270338      0.981227      1.264080   \n",
       "BIOMASS__1            0.000000      0.000000      0.000000      0.000000   \n",
       "ORNTA                      inf           inf           inf           inf   \n",
       "ATPM                  0.000000      0.000000      0.000000      0.000000   \n",
       "ACOATA                1.510789      1.704992      1.232206      1.550081   \n",
       "5DOAN                 0.000000      0.000000      0.000000      0.000000   \n",
       "3HAD160               1.139276      1.271588      0.927577      1.107939   \n",
       "RNDR3                 2.981074      2.837915      2.356418      2.085642   \n",
       "Htex                  0.000000      0.000000      0.000000      0.000000   \n",
       "G5SD                  1.040765      1.169482      1.002883      1.290664   \n",
       "HISTDa                0.968840      1.089301      0.961113      1.256080   \n",
       "TRPS3                 1.037566      1.190476      0.894709      1.110053   \n",
       "DMBZIDS2              1.492613      1.613869      1.365633      1.533209   \n",
       "HSTPT                 1.466724      1.496350      1.341262      1.486818   \n",
       "\n",
       "                  Se_axen_d6_2  Se_axen_d6_3  Se_axen_d8_1  Se_axen_d8_2  \\\n",
       "Reaction_ID                                                                \n",
       "EX_gln__L_e           0.000000      0.000000      0.000000      0.000000   \n",
       "EX_hco3_e             0.000000      0.000000      0.000000      0.000000   \n",
       "EX_mn2_e              0.000000      0.000000      0.000000      0.000000   \n",
       "EX_arg__L_e           0.000000      0.000000      0.000000      0.000000   \n",
       "ADPT                       inf           inf           inf           inf   \n",
       "O2tcx                 0.000000      0.000000      0.000000      0.000000   \n",
       "AOXPBDC               0.000000      0.000000      0.000000      0.000000   \n",
       "DNTPPA                0.000000      0.000000      0.000000      0.000000   \n",
       "CBMD                  0.000000      0.000000      0.000000      0.000000   \n",
       "BIOMASS_PIGMENTS      0.000000      0.000000      0.000000      0.000000   \n",
       "H2Otu_syn             0.000000      0.000000      0.000000      0.000000   \n",
       "DTMPK                 1.100125      1.086358      0.591990      0.861076   \n",
       "BIOMASS__1            0.000000      0.000000      0.000000      0.000000   \n",
       "ORNTA                      inf           inf           inf           inf   \n",
       "ATPM                  0.000000      0.000000      0.000000      0.000000   \n",
       "ACOATA                1.518841      1.305636      0.576490      1.134944   \n",
       "5DOAN                 0.000000      0.000000      0.000000      0.000000   \n",
       "3HAD160               1.027855      0.695682      0.495125      0.788301   \n",
       "RNDR3                 1.883802      1.591246      1.324064      1.152886   \n",
       "Htex                  0.000000      0.000000      0.000000      0.000000   \n",
       "G5SD                  1.013909      1.038590      0.627251      0.857526   \n",
       "HISTDa                1.042307      1.101714      0.633719      0.972682   \n",
       "TRPS3                 0.948148      0.962434      0.610582      0.901323   \n",
       "DMBZIDS2              1.353254      1.249434      0.786903      1.091308   \n",
       "HSTPT                 1.224216      1.152769      0.735595      1.011078   \n",
       "\n",
       "                  Se_axen_d8_3  \n",
       "Reaction_ID                     \n",
       "EX_gln__L_e                0.0  \n",
       "EX_hco3_e                  0.0  \n",
       "EX_mn2_e                   0.0  \n",
       "EX_arg__L_e                0.0  \n",
       "ADPT                       inf  \n",
       "O2tcx                      0.0  \n",
       "AOXPBDC                    0.0  \n",
       "DNTPPA                     0.0  \n",
       "CBMD                       0.0  \n",
       "BIOMASS_PIGMENTS           0.0  \n",
       "H2Otu_syn                  0.0  \n",
       "DTMPK                      1.0  \n",
       "BIOMASS__1                 0.0  \n",
       "ORNTA                      inf  \n",
       "ATPM                       0.0  \n",
       "ACOATA                     1.0  \n",
       "5DOAN                      0.0  \n",
       "3HAD160                    1.0  \n",
       "RNDR3                      1.0  \n",
       "Htex                       0.0  \n",
       "G5SD                       1.0  \n",
       "HISTDa                     1.0  \n",
       "TRPS3                      1.0  \n",
       "DMBZIDS2                   1.0  \n",
       "HSTPT                      1.0  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create dataframe of normalized enzyme activity by scaling realtive to reference strain\n",
    "enzymze_activity_df = all_enzyme_activity_df.copy()\n",
    "\n",
    "rep_names = transcriptomics_df.columns\n",
    "normalized_enzyme_activity_df = enzymze_activity_df[rep_names].div(enzymze_activity_df[ref_rep], axis=0)\n",
    "\n",
    "\n",
    "# TODO: should these replacements be the original values (0 and inf), or 1s (to indicate same as reference strain)?\n",
    "# ==> NEITHER: drop rows with all 0 or inf or NaN entries\n",
    "\n",
    "# Ensure 0-valued or inf entries in enzymze_activity_df are passed into normalized_enzyme_activity_df\n",
    "# - OR - \n",
    "# Ensure 0-valued or inf entries in enzymze_activity_df are passed into normalized_enzyme_activity_df\n",
    "for col in rep_names:\n",
    "    for row in enzymze_activity_df.index:\n",
    "        if (enzymze_activity_df[col][row] == 0.0) or (np.isinf(enzymze_activity_df[col][row])):\n",
    "            # this_row = enzymze_activity_df['Reaction_ID'][row]\n",
    "            # normalized_enzyme_activity_df.loc[this_row, col] = enzymze_activity_df[col][row]\n",
    "            normalized_enzyme_activity_df.loc[row, col] = enzymze_activity_df[col][row]\n",
    "\n",
    "# Re-index normalized_enzyme_activity_df using reaction-IDs\n",
    "normalized_enzyme_activity_df['Reaction_ID'] = enzymze_activity_df['Reaction_ID']\n",
    "normalized_enzyme_activity_df = normalized_enzyme_activity_df.set_index('Reaction_ID')\n",
    "\n",
    "normalized_enzyme_activity_df.iloc[0:25]\n",
    "\n",
    "# TODO: add \"check_data\" function to check validity before passing to pymc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BIOMASS__1': False, 'EX_co2_e': False, 'EX_sucr_e': False}"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "missing_rxns = {r: (r in FVA_object.index) for r in normalized_enzyme_activity_df.index}\n",
    "{k:v for k,v in missing_rxns.items() if not v}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ----- STOP HERE ..... for now"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4. For each condition, multiply the reference bounds by the normalized enzyme data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get relative scaled rates of change in metabolic abundance\n",
    "normalized_rates_df = rates_df.div(rates_df[ref_rep], axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>d4_1_min</th>\n",
       "      <th>d4_1_max</th>\n",
       "      <th>minimum</th>\n",
       "      <th>maximum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>EX_gln__L_e</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_hco3_e</th>\n",
       "      <td>-1.990000e+00</td>\n",
       "      <td>-1.990000e+00</td>\n",
       "      <td>-1.990000e+00</td>\n",
       "      <td>-1.990000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_mn2_e</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_arg__L_e</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADPT</th>\n",
       "      <td>1.032064e-05</td>\n",
       "      <td>1.032064e-05</td>\n",
       "      <td>1.032064e-05</td>\n",
       "      <td>1.032064e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O2tcx</th>\n",
       "      <td>2.631984e-02</td>\n",
       "      <td>2.631984e-02</td>\n",
       "      <td>2.631984e-02</td>\n",
       "      <td>2.631984e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AOXPBDC</th>\n",
       "      <td>1.411151e-06</td>\n",
       "      <td>1.411151e-06</td>\n",
       "      <td>1.411151e-06</td>\n",
       "      <td>1.411151e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DNTPPA</th>\n",
       "      <td>4.111581e-06</td>\n",
       "      <td>4.111581e-06</td>\n",
       "      <td>4.111581e-06</td>\n",
       "      <td>4.111581e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CBMD</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BIOMASS_PIGMENTS</th>\n",
       "      <td>1.061867e-03</td>\n",
       "      <td>1.061867e-03</td>\n",
       "      <td>1.061867e-03</td>\n",
       "      <td>1.061867e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H2Otu_syn</th>\n",
       "      <td>-7.837161e+00</td>\n",
       "      <td>-7.837161e+00</td>\n",
       "      <td>-7.837161e+00</td>\n",
       "      <td>-7.837161e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DTMPK</th>\n",
       "      <td>2.847641e-04</td>\n",
       "      <td>2.847641e-04</td>\n",
       "      <td>2.847641e-04</td>\n",
       "      <td>2.847641e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORNTA</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATPM</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACOATA</th>\n",
       "      <td>2.399223e-02</td>\n",
       "      <td>2.399223e-02</td>\n",
       "      <td>2.399223e-02</td>\n",
       "      <td>2.399223e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5DOAN</th>\n",
       "      <td>5.599703e-06</td>\n",
       "      <td>5.599703e-06</td>\n",
       "      <td>5.599703e-06</td>\n",
       "      <td>5.599703e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3HAD160</th>\n",
       "      <td>2.296164e-02</td>\n",
       "      <td>2.296164e-02</td>\n",
       "      <td>2.296164e-02</td>\n",
       "      <td>2.296164e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNDR3</th>\n",
       "      <td>6.388443e-04</td>\n",
       "      <td>6.388443e-04</td>\n",
       "      <td>6.388443e-04</td>\n",
       "      <td>6.388443e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Htex</th>\n",
       "      <td>2.427223e+00</td>\n",
       "      <td>2.427223e+00</td>\n",
       "      <td>2.427223e+00</td>\n",
       "      <td>2.427223e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G5SD</th>\n",
       "      <td>1.275780e-02</td>\n",
       "      <td>1.275780e-02</td>\n",
       "      <td>1.275780e-02</td>\n",
       "      <td>1.275780e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HISTDa</th>\n",
       "      <td>4.626423e-03</td>\n",
       "      <td>4.626423e-03</td>\n",
       "      <td>4.626423e-03</td>\n",
       "      <td>4.626423e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TRPS3</th>\n",
       "      <td>3.852138e-03</td>\n",
       "      <td>3.852138e-03</td>\n",
       "      <td>3.852138e-03</td>\n",
       "      <td>3.852138e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DMBZIDS2</th>\n",
       "      <td>2.758159e-07</td>\n",
       "      <td>2.758159e-07</td>\n",
       "      <td>2.758159e-07</td>\n",
       "      <td>2.758159e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSTPT</th>\n",
       "      <td>4.626423e-03</td>\n",
       "      <td>4.626423e-03</td>\n",
       "      <td>4.626423e-03</td>\n",
       "      <td>4.626423e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCAROTDH2</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORNDC</th>\n",
       "      <td>6.927468e-06</td>\n",
       "      <td>6.927468e-06</td>\n",
       "      <td>6.927468e-06</td>\n",
       "      <td>6.927468e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GND</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMPA_1</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>U23GAAT2</th>\n",
       "      <td>5.145925e-04</td>\n",
       "      <td>5.145925e-04</td>\n",
       "      <td>5.145925e-04</td>\n",
       "      <td>5.145925e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HOXGfx</th>\n",
       "      <td>5.564182e-05</td>\n",
       "      <td>5.564182e-05</td>\n",
       "      <td>5.564182e-05</td>\n",
       "      <td>5.564182e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGK</th>\n",
       "      <td>4.638910e+00</td>\n",
       "      <td>4.638910e+00</td>\n",
       "      <td>4.638910e+00</td>\n",
       "      <td>4.638910e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IMPD</th>\n",
       "      <td>5.745098e-03</td>\n",
       "      <td>5.745098e-03</td>\n",
       "      <td>5.745098e-03</td>\n",
       "      <td>5.745098e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LPOR</th>\n",
       "      <td>6.043084e-04</td>\n",
       "      <td>6.043084e-04</td>\n",
       "      <td>6.043084e-04</td>\n",
       "      <td>6.043084e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CHRPL</th>\n",
       "      <td>9.621483e-07</td>\n",
       "      <td>9.621483e-07</td>\n",
       "      <td>9.621483e-07</td>\n",
       "      <td>9.621483e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCALDDy</th>\n",
       "      <td>4.111581e-06</td>\n",
       "      <td>4.111581e-06</td>\n",
       "      <td>4.111581e-06</td>\n",
       "      <td>4.111581e-06</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      d4_1_min      d4_1_max       minimum       maximum\n",
       "EX_gln__L_e       0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00\n",
       "EX_hco3_e        -1.990000e+00 -1.990000e+00 -1.990000e+00 -1.990000e+00\n",
       "EX_mn2_e          0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00\n",
       "EX_arg__L_e       0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00\n",
       "ADPT              1.032064e-05  1.032064e-05  1.032064e-05  1.032064e-05\n",
       "O2tcx             2.631984e-02  2.631984e-02  2.631984e-02  2.631984e-02\n",
       "AOXPBDC           1.411151e-06  1.411151e-06  1.411151e-06  1.411151e-06\n",
       "DNTPPA            4.111581e-06  4.111581e-06  4.111581e-06  4.111581e-06\n",
       "CBMD              0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00\n",
       "BIOMASS_PIGMENTS  1.061867e-03  1.061867e-03  1.061867e-03  1.061867e-03\n",
       "H2Otu_syn        -7.837161e+00 -7.837161e+00 -7.837161e+00 -7.837161e+00\n",
       "DTMPK             2.847641e-04  2.847641e-04  2.847641e-04  2.847641e-04\n",
       "ORNTA             0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00\n",
       "ATPM              0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00\n",
       "ACOATA            2.399223e-02  2.399223e-02  2.399223e-02  2.399223e-02\n",
       "5DOAN             5.599703e-06  5.599703e-06  5.599703e-06  5.599703e-06\n",
       "3HAD160           2.296164e-02  2.296164e-02  2.296164e-02  2.296164e-02\n",
       "RNDR3             6.388443e-04  6.388443e-04  6.388443e-04  6.388443e-04\n",
       "Htex              2.427223e+00  2.427223e+00  2.427223e+00  2.427223e+00\n",
       "G5SD              1.275780e-02  1.275780e-02  1.275780e-02  1.275780e-02\n",
       "HISTDa            4.626423e-03  4.626423e-03  4.626423e-03  4.626423e-03\n",
       "TRPS3             3.852138e-03  3.852138e-03  3.852138e-03  3.852138e-03\n",
       "DMBZIDS2          2.758159e-07  2.758159e-07  2.758159e-07  2.758159e-07\n",
       "HSTPT             4.626423e-03  4.626423e-03  4.626423e-03  4.626423e-03\n",
       "ZCAROTDH2         0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00\n",
       "ORNDC             6.927468e-06  6.927468e-06  6.927468e-06  6.927468e-06\n",
       "GND               0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00\n",
       "CMPA_1            0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00\n",
       "U23GAAT2          5.145925e-04  5.145925e-04  5.145925e-04  5.145925e-04\n",
       "HOXGfx            5.564182e-05  5.564182e-05  5.564182e-05  5.564182e-05\n",
       "PGK               4.638910e+00  4.638910e+00  4.638910e+00  4.638910e+00\n",
       "IMPD              5.745098e-03  5.745098e-03  5.745098e-03  5.745098e-03\n",
       "LPOR              6.043084e-04  6.043084e-04  6.043084e-04  6.043084e-04\n",
       "CHRPL             9.621483e-07  9.621483e-07  9.621483e-07  9.621483e-07\n",
       "GCALDDy           4.111581e-06  4.111581e-06  4.111581e-06  4.111581e-06"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TODO: make this into a function\n",
    "flux_bounds = {}\n",
    "for this_strain in normalized_enzyme_activity_df.columns:\n",
    "    flux_bounds[this_strain] = FVA_object\n",
    "    for this_rxn in normalized_enzyme_activity_df.index:\n",
    "        if this_rxn in FVA_object:\n",
    "            this_factor = normalized_enzyme_activity_df[this_strain][this_rxn]\n",
    "            flux_bounds[this_strain].loc[this_rxn] = this_factor*FVA_object[this_strain][this_rxn]\n",
    "        elif this_rxn in missing_rxns:\n",
    "            # TODO: for sucrose production ('EX_sucr_e'), use normalized calcualted rate from normalized_rates_df\n",
    "            if rxn=='EX_sucr_e':\n",
    "                # flux_bounds[this_strain].loc[this_rxn] = pd.Series({'minimum': ....,\n",
    "                #                                                     'maximum': ....})\n",
    "                \n",
    "\n",
    "test_df = flux_bounds['Se_axen_d4_1'].iloc[0:35]\n",
    "test_df.columns = ['d4_1_min', 'd4_1_max']\n",
    "test_df.join(flux_bounds[ref_rep].iloc[0:35])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "minimum    0.0\n",
       "maximum    0.0\n",
       "Name: EX_gln__L_e, dtype: float64"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "FVA_object.iloc[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5. Run FBA using the condition-specific bounds to compute the condition-specific fluxes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bmca",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
